src/server/pkg/storage/storage-format-spec.md

```diff
@@ -128,42 +132,135 @@ important differences between these two types of objects as it pertains to GC is
-describes all of the referenced blocks in our cluster, we can keep this up
-to date by adding the bloom filters of newly added blocks, and subtracting
-the bloom filters of deleted blocks. One of the difficulties with bloom
-filters is that they can't be used to enumerate paths, they can only be
-used to check a path that you already know about. That means that to
-perform garbage collection we'll need to scan objects and check them
-against our bloom filter to see if they're still referenced. I think this
-can work fine as a process that is always passively running.
-
-The other big difficulty with Pachyderm GC is race conditions. Because of
-our deduplication system, an object that is unreferenced and thus ready to be
-GCed may at any point get re-referenced and thus cease to be ready for GC.
-I think we can solve this with a generational system wherein each object
-belongs to a generation, and we record a bloom filter that describes the
-reference changes in a generation. By summing all the generations together
-we can get a global bloom filter for a generation and figure out what's
-safe to delete. This is best demonstrated with an example:
-
-| gen | ops          | bloom     | bloom sum |
-| --- | ------------ | ------    | --------- |
-|   0 | +foo         | 100101    | 100101    |
-|   1 | +bar         | 001011    | 101112    |
-|   2 | -foo         | -100-10-1 | 001011    |
-
-After generation 2, foo is ready to be deleted, which is reflected in the
-fact that the bloom sum returns negative for foo's bloom filter (100101).
-The tricky part is that if we now want to re-add foo in gen 3 we'll need
-to make sure that the GC process for gen 2 has either already finished,
-because if it hasn't, it might GC foo after we write it for gen 3, or
-knows not to GC this object, because it's about to be re-referenced.
-I think all of this is manageable, but it's definitely tricky.
+### Requirements
+
+* Always-running background process/service
+* No global locks or stoppages required during steady-state operations
+* Iterative and rate-limitable for any large operations
+* Batches updates from Pachyderm and batches deletes to object storage
+* Recovers smoothly from disorderly shutdown
+* Capable of becoming shardable with minimal changes
+
+### API
+
+The garbage collector will have two components, a client-side library and a
+service that runs in pachd.  The library will act as the single-point of contact
+for a client, while coordinating between the pachd service and the database as
+necessary.
+
+```go
+func (gcc *GcClient) ReserveChunks(chunks []Chunk)
+func (gcc *GcClient) UpdateReferences(add []Chunk, remove []Chunk)
+```
+
+The garbage collection service will provide a GRPC API for use by the client
+library (which should not be used directly).
+
+```
+rpc FlushDeletes(FlushDeletesRequest)
+```
+
+When deleting a chunk, the garbage collection service will need to evaluate the
```

Comment:
This would essentially amount to walking a FileSet (index) and making `UpdateReferences` calls. I wonder if it makes more sense to not think about the transitive closure of a chunk (as the chunk layer / garbage collection sees it) and instead think in terms of FileSet deletions going through another part of the storage layer which makes a sequence of `UpdateReferences` calls. I don't think there is a case where we would delete a chunk without some notion of deleting / modifying a FileSet. The main scenario we would need to think about is a crash/failure on the client side in the middle of a set of deletions. At a high level, it seems like we would need to persist an identifier associated with a set of operations (maybe the semantic name of an object being created/modified/deleted) and the most recent chunk reference (the `UpdateReferences` operation could happen in a transaction and the last modified would be stored). Also, we/I need to think about what having multiple files in a chunk means in terms of reference counts (increment reference count for each file in a chunk or one increment for the chunk).

Reply:
Also, a good reason not to have the garbage collector resolve the transitive closure of an object is simply because garbage collection could quickly become a bottleneck. This is a global system, so if we can push more work on the workers then I think, in general, that is better.

Reply:
> I don't think there is a case where we would delete a chunk without some notion of deleting / modifying a FileSet.
Would Index merging/compaction not potentially delete old chunks?

> The main scenario we would need to think about is a crash/failure on the client side in the middle of a set of deletions.
I think this may be out of the scope of a garbage collector.  If the file layer performs a partial deletion of some number of semantic references, it is the file layer's responsibility to persist its state and resume.  The garbage collector should just ensure that unreferenced chunks are eventually removed and referenced chunks are never removed.

Currently, the plan is for a set of `UpdateReferences` to happen transactionally to ensure nothing gets accidentally removed while in an intermediate state.

> This is a global system, so if we can push more work on the workers then I think, in general, that is better.
I believe this design could be effectively sharded by having multiple garbage collector services, each responsible for a prefix range of the chunk keyspace.  I don't think I'm advocating for resolving transitive closures of references, though (see my earlier comment).

Reply:
> Would Index merging/compaction not potentially delete old chunks?

Yes, but that is still happening through FileSet abstractions.

> I think this may be out of the scope of a garbage collector. If the file layer performs a partial deletion of some number of semantic references, it is the file layer's responsibility to persist its state and resume. The garbage collector should just ensure that unreferenced chunks are eventually removed and referenced chunks are never removed.

> Currently, the plan is for a set of UpdateReferences to happen transactionally to ensure nothing gets accidentally removed while in an intermediate state.

Maybe. It seems like it would fit well under the garbage collection umbrella because it is an operation interfacing with the garbage collector, but I don't think I am against having it be just some other table that workers interface with.

> I believe this design could be effectively sharded by having multiple garbage collector services, each responsible for a prefix range of the chunk keyspace. I don't think I'm advocating for resolving transitive closures of references, though (see my earlier comment).

Scaling up the number of workers is going to be a lot easier/more performant than scaling up the number of garbage collector shards. We should do as much work outside of the garbage collector as we can.